scipy/stats/stats.py

```diff
@@ -2104,28 +2116,90 @@ def percentileofscore(a, score, kind='rank'):
+    array([40., 70.])
+
+    The inputs can be infinite:
+
+    >>> stats.percentileofscore([-np.inf, 0, 1, np.inf], [1, 100])
+    array([75., 75.])
+
+    If `a` is empty, or containing only nan's (which are omitted),
+    the resulting percentiles are all nan:
+
+    >>> stats.percentileofscore([], 1)
+    nan
+    >>> stats.percentileofscore([np.nan, 1, 2], [1, 2], nan_policy='omit')
+    array([ 50., 100.])
+    >>> stats.percentileofscore([np.nan, np.nan], [1, 2], nan_policy='omit')
+    array([nan, nan])
+
+    The 'omit' policy does not make sense if there are nan's among the scores,
+    and so an error in that case.
     """
-    if np.isnan(score):
-        return np.nan
+
     a = np.asarray(a)
     n = len(a)
+    score = np.asarray(score)
+
+    # Nan treatment
+    cna, npa = _contains_nan(a, nan_policy, summary_check=False)
+    cns, nps = _contains_nan(score, nan_policy, summary_check=False)
+
+    if (cna or cns) and nan_policy == 'raise':
+        raise ValueError("The input contains nan values")
+
+    elif cns and nan_policy == "omit":
+        raise ValueError("The input scores contains nan values,"
+                         " which is incompatible with the 'omit' policy.")
+
+    # If a score is nan, then the output should be nan
+    if cns and nan_policy == "propagate":
+        score = ma.masked_where(np.isnan(score), score)
+
+    # Don't count nans
+    if cna and nan_policy == "omit":
+        a = ma.masked_where(np.isnan(a), a)
+        n = a.count()
+
+    # With `propagate`, an `a` with nans should always yield nan
+    if cna and nan_policy == "propagate":
+        n = 0
+
+    # Cannot compare to empty list ==> nan
     if n == 0:
-        return 100.0
-
-    if kind == 'rank':
-        left = np.count_nonzero(a < score)
-        right = np.count_nonzero(a <= score)
-        pct = (right + left + (1 if right > left else 0)) * 50.0/n
-        return pct
-    elif kind == 'strict':
-        return np.count_nonzero(a < score) / n * 100
-    elif kind == 'weak':
-        return np.count_nonzero(a <= score) / n * 100
-    elif kind == 'mean':
-        pct = (np.count_nonzero(a < score) + np.count_nonzero(a <= score)) / n * 50
-        return pct
+        perct = np.full_like(score, np.nan, dtype=float)
+
     else:
-        raise ValueError("kind can only be 'rank', 'strict', 'weak' or 'mean'")
+        # Prepare broadcasting
+        score = score[..., None]
+        count = lambda x: np.count_nonzero(x, -1)
+
+        # Compute
+        if kind == 'rank':
+            left  = count(a < score)
+            right = count(a <= score)
+            plus1 = left < right
+            perct = (left + right + plus1) * (50.0 / n)
+        elif kind == 'strict':
+            perct = count(a < score) * (100.0 / n)
+        elif kind == 'weak':
+            perct = count(a <= score) * (100.0 / n)
+        elif kind == 'mean':
+            left  = count(a < score)
+            right = count(a <= score)
```

Comment:
Seems like you would need to use `numpy.ma.masked_less` instead!

Reply:
Thanks! That's interesting. Indeed, I believe it could be worked using `np.ma.masked_less` & co. However, it seems like there would be an inefficiency in that: these routines create a new `MaskedArray`, whereas `a < score` simply creates the boolean mask.

Furthermore, although I'm not well versed with masked arrays, the error message produced on Azure pipeline (presumably complaining about nan's) makes little sense to me. Operations on `MaskedArray`s are not really supposed to process the masked elements. Furthermore, the error does not seem to occur on Linux/Mac platforms, adding to my suspicion that something funny is going on. I note that the error is a `RunTimeWarning`, so maybe it could just be silenced?

Reply:
I never worked with masked array. Maybe ask another maintainer here.

But sure, you could try to silent the warning and see if tests pass.

Reply:
Seems like that did it 🥳

Reply:
GG. We just need to have another eye on this as I have no experience with masked arrays.

Reply:
@patricknraanes did you experiment at all with masking the arrays yourself (e.g. `x = x[~np.isnan(x)]`) rather than using masked arrays?

Reply:
As I recall, I think I did. I don't remember very well, but I think I chose the `masked_array` approach because it made it easier to support all cases of `nan_policy`.

Reply:
Ok. Probably fine. I haven't looked closely about which of the changes are related to masked array and which are due to array second argument. If I see major simplifications that would result from manual masking, i may suggest them. (Masked arrays tend to be slower in my experience, so they would have to make the code simpler to justify them.)

Reply:
If we decide that vectorizing this only over `scores` and implementing `nan_policy='omit'` only for `a` is OK, I will try an implementation that avoids masked arrays and see which looks simpler and which is faster, mostly because I'm curious.